server/src/com/google/refine/Refine.java

```diff
@@ -494,12 +494,42 @@ public void actionPerformed(ActionEvent e) {
     
     private void openBrowser() {
         if (!Desktop.isDesktopSupported()) {
-            logger.warn("Java Desktop class not supported on this platform.  Please open %s in your browser",uri.toString());
+            try {
+                openBrowser2();
+            } catch (IOException e) {
+                logger.error("Failed to open browser. Please open %s in your browser",uri.toString());
+            }
+        } else {
+            try {
+                Desktop.getDesktop().browse(uri);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
         }
-        try {
-            Desktop.getDesktop().browse(uri);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+    }
+
+    private void openBrowser2() throws IOException {
+        logger.info("Opening browser to {}", uri.toString());
+        String os = System.getProperty("os.name").toLowerCase();
+        Runtime runtime = Runtime.getRuntime();
+        if(os.contains("win")) { // If we are on Windows
+            runtime.exec("rundll32 url.dll,FileProtocolHandler " + uri);
+        } else if(os.contains("mac")) { // If we are on Mac
+            runtime.exec("open " + uri);
+        } else if(os.contains("nix") || os.contains("nux")) { // If we are on unix
+            String[] browsers = {"google-chrome", "epiphany", "firefox", "mozilla", "konqueror",
+                    "netscape", "opera", "links", "lynx" };
```

Comment:
Perhaps that is a bit much… Also, it will try browsers in this exact order, whereas we might want to open the URL with the default browser instead. How about using `xdg-open` (and potentially its alternatives) instead?

Reply:
```
import java.awt.Desktop;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

public class MultiBrowPop {

    public static void main(String[] args) {
        OUT("\nWelcome to Multi Brow Pop.\nThis aims to popup a browsers in multiple operating systems.\nGood luck!\n");

        String url = "http://www.birdfolk.co.uk/cricmob";
        OUT("We're going to this page: "+ url);

        String myOS = System.getProperty("os.name").toLowerCase();
        OUT("(Your operating system is: "+ myOS +")\n");

        try {
            if(Desktop.isDesktopSupported()) { // Probably Windows
                OUT(" -- Going with Desktop.browse ...");
                Desktop desktop = Desktop.getDesktop();
                desktop.browse(new URI(url));
            } else { // Definitely Non-windows
                Runtime runtime = Runtime.getRuntime();
                if(myOS.contains("mac")) { // Apples
                    OUT(" -- Going on Apple with 'open'...");
                    runtime.exec("open " + url);
                } 
                else if(myOS.contains("nix") || myOS.contains("nux")) { // Linux flavours 
                    OUT(" -- Going on Linux with 'xdg-open'...");
                    runtime.exec("xdg-open " + url);
                }
                else 
                    OUT("I was unable/unwilling to launch a browser in your OS :( #SadFace");
            }
            OUT("\nThings have finished.\nI hope you're OK.");
        }
        catch(IOException | URISyntaxException eek) {
            OUT("**Stuff wrongly: "+ eek.getMessage());
        }
    }
```
Will this be ok to use @wetneb as it is using xdg-open?

Reply:
It looks pretty good to me indeed, thank you!